src/pandas/array.h

```diff
@@ -42,10 +52,7 @@ class Array {
   virtual bool owns_data() const = 0;
 
  protected:
-  std::shared_ptr<DataType> type_;
-  int64_t length_;
```

Comment:
I think the CategoryArray is the counterexample. You need to keep the CategoryArray's codes_ member's length in sync with the Array's length. That seems very error prone. There might be an intermediate template class that is used by most Array sub-classes excep the CategoryArray:

```cpp
template <typename DATA_TYPE>
class ArrayImpl
{
    protected:
        std::shared_ptr<DATA_TYPE) type_;
        int64_t length;  
};
```